### PR TITLE
Adding bounded-time start of SamzaContainer from StreamProcessor

### PR DESCRIPTION
**Pull request by rmatharu**
This bounded-time start is needed to account for cases where the start of a container never returns within a bounded amount of time, e.g., when querying metadata starts taking a long amount of time. 
We would like the container to exit gracefully in that case.

### CODE REVIEW

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java [hunk 1]
```diff
@@ -401,11 +403,29 @@ public void onJobModelExpired() {
...
+                LOGGER.error("Timed out {} ms while waiting for the container: {} to startup. Interrupting the container: {} thread to die", taskStartMs, container, container);
+                containerExcecutorService.shutdownNow();
+                state = State.STOPPED;
+                jobCoordinator.stop();
+                processorListener.afterFailure(new TimeoutException("Container start timed out after "+taskStartMs+ " ms"));
+              }
+            } catch (InterruptedException e) {
+              LOGGER.error("Exception occurred when starting the container: {}.", container, e);
+              state = State.STOPPED;
+              jobCoordinator.stop();
```
Comment (Sanil15) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java]: Shouldn't this be in finally?

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java [hunk 2]
```diff
@@ -401,11 +403,29 @@ public void onJobModelExpired() {
...
             container = createSamzaContainer(processorId, jobModel);
             container.setContainerListener(new ContainerListener());
             LOGGER.info("Starting the container: {} for the stream processor: {}.", container, processorId);
             containerExcecutorService.submit(container);
+
+            try {
+              LOGGER.info("Waiting {} ms for the container: {} to startup.", taskStartMs, container);
+
+              if (!containerStartLatch.await(taskStartMs, TimeUnit.MILLISECONDS)) {
+                LOGGER.error("Timed out {} ms while waiting for the container: {} to startup. Interrupting the container: {} thread to die", taskStartMs, container, container);
```
Comment (Sanil15) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java]: nit: "Timed out after {} ms, waiting for the container ...."

File: samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java [hunk 3]
```diff
@@ -401,11 +403,29 @@ public void onJobModelExpired() {
...
         synchronized (lock) {
           if (state == State.IN_REBALANCE) {
+            containerStartLatch = new CountDownLatch(1);
             containerShutdownLatch = new CountDownLatch(1);
             container = createSamzaContainer(processorId, jobModel);
             container.setContainerListener(new ContainerListener());
             LOGGER.info("Starting the container: {} for the stream processor: {}.", container, processorId);
             containerExcecutorService.submit(container);
+
+            try {
```
Comment (Sanil15) [inline samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java]: Consider this: Move try inside if and then move:

`state = State.STOPPED;
jobCoordinator.stop();
processorListener.afterFailure(new TimeoutException("Container start timed out after "+taskStartMs+ " ms"));`

inside a finally block

**Comment by rmatharu** [comment]
Closing this since
deciding the bound on the start of a container is non-trivial, and depends on a large number of parameters. 
Instead, we can introduce bounded-start on components such as SystemsConsumer.

**Pull request closed without merging**